src/transience.rs

```diff
@@ -122,24 +127,21 @@ use std::marker::PhantomData;
 /// For example, a type `&'long i32` with a transience of `Co<'long>` (middle row)
 /// can be safely erased to the trait object `dyn Any<Co<'short>>` (3rd column) by
 /// shortening its lifetime. However, when we want to
-/// [`downcast`] the opaque trait object back into its concrete type, we _cannot_ just
-/// assume (at least under the implementation of this crate) that we can safely recover
-/// a `'long` lifetime from the `'short` lifetime of the trait object;
-/// instead, we can only safely recover `&'short i32` from the `dyn Any<Co<'short>>`.
-/// As a result, even though the underlying reference should actually be valid for the
-/// `'long` lifetime, we end up with a reference that the borrow-checker only believes
-/// to be valid for the `'short` lifetime. This may seem silly if you happen to be
-/// dealing with a case where _you know_ what you are working with and that the
-/// `'long` lifetime _is_ valid; but the whole point of type-erasure is that
-/// sometimes you _dont_ know/care, or you don't want to compiler to know/care,
-/// and sometimes ignorance has a cost. On the bright side, if you _really are_
-/// confident that `'long` is valid, you can `unsafe`-ly fix it using hacks like
-/// [`std::mem::transmute`] and raw pointer casts (at your own risk, of course).
+/// [`downcast`] the opaque trait object back into its concrete type, we _cannot_
+/// assume that we can safely recover
+/// a `'long` lifetime from the `'short` lifetime of the trait object. The only truly
+/// safe solution is to avoid downcasting to a shorter lifetime - just as if you were
+/// working with references to the type itself.
 ///
 /// [`Transience` associated type]: Transient::Transience
 /// [variance]: https://doc.rust-lang.org/nomicon/subtyping.html
 /// [`downcast`]: crate::Downcast::downcast
-pub trait Transience: Transient + CanTranscendTo<Self> + CanRecoverFrom<Self> {}
+pub trait Transience: Transient + CanTranscendTo<Self> + CanRecoverFrom<Self> + Sealed {
```

Comment:
I agree that the new associated types make this trait safety critical and it needs to be protected somehow, but I'm torn on whether the answer is sealing it or just marking it `unsafe`. I was leaning towards taking the `unsafe` approach since there are limitations to set of `Transience` impls this crate provides, and I can image a world where someone would want to implement a custom transience for their unsupported type. Is there a reason in particular that you'd prefer sealing it?

Reply:
My main motivation was protecting it (since it's safety critical), and my choice was somewhat arbitrary. 

That being said, the other argument in favor of sealing is that there aren't really any other valid types. However, while thinking it through, I'm not convinced that it's 100% impossible to wind up with something weird, and leaving it unsafe (but not sealed) is probably a good move.